lib/burnside/intrinsics.cc

```diff
@@ -447,31 +450,23 @@ mlir::Value *IntrinsicLibrary::Implementation::generateRuntimeCall(
 }
 
 // CONJG
-mlir::Value *IntrinsicLibrary::Implementation::conjgGenerator(
+mlir::Value *IntrinsicLibrary::Implementation::generateConjg(
     Context &genCtxt) const {
   assert(genCtxt.arguments.size() == 1);
   mlir::Type resType{genCtxt.getResultType()};
   assert(resType == genCtxt.arguments[0]->getType());
+  mlir::OpBuilder &builder{*genCtxt.builder};
+  ComplexHandler cplxHandler{builder, genCtxt.loc};
 
-  // FIXME make fir.complex<f32>.. instead of fir.complex<4> to avoid
-  // constantly having to translate to fe types ? Or get actual mapping
-  auto complexType{resType.cast<fir::CplxType>()};
-  mlir::Type realType{
-      burnside::convertReal(genCtxt.getMLIRContext(), complexType.getFKind())};
-  auto ity{mlir::IndexType::get(genCtxt.getMLIRContext())};
-  auto oneAttr{genCtxt.builder->getIntegerAttr(ity, 1)};
-  auto *imagIdx{
-      genCtxt.builder->create<mlir::ConstantOp>(genCtxt.loc, ity, oneAttr)
-          .getResult()};
-  auto parts{genCtxt.builder->create<fir::ExtractValueOp>(
-      genCtxt.loc, realType, genCtxt.arguments[0], imagIdx)};
   // TODO a negation unary would be better than a sub to zero ?
-  auto zeroAttr{genCtxt.builder->getFloatAttr(realType, llvm::APFloat{0.})};
-  auto zero{genCtxt.builder->create<mlir::ConstantOp>(
-      genCtxt.loc, realType, zeroAttr)};
-  auto negImag{genCtxt.builder->create<mlir::SubFOp>(genCtxt.loc, zero, parts)};
-  return genCtxt.builder->create<fir::InsertValueOp>(
-      genCtxt.loc, resType, genCtxt.arguments[0], negImag, imagIdx);
+  mlir::Value *cplx{genCtxt.arguments[0]};
+  mlir::Type realType{cplxHandler.getComplexPartType(cplx)};
+  mlir::Value *zero{builder.create<mlir::ConstantOp>(
+      genCtxt.loc, realType, builder.getZeroAttr(realType))};
+  mlir::Value *imag{cplxHandler.createComplexImagPart(cplx)};
+  mlir::Value *negImag{
```

Comment:
Don't you want to use `auto` and `auto *` throughout?  It will simplify refactoring in the future.

Reply:
I am torn apart here. I know llvm is ok with this and f18 is not.
I agree that the code here is mostly splitting what could have been chained in an awful single function call where I would not have spelled out the types.

But I find `auto` is not very compassionate for new code readers. It assume he or she is familiar enough with the functions used to guess the types. I think knowing the types helps a lot understanding what is happening.

Also, I am completely sure it is safe regarding refactoring that would do more than renaming, I am afraid it may do weird things combined with C++ implicit casts in function arguments that would not be caught at compile time sometimes.

So for small core types such as `mlir::Type` and `mlir::Value*`, I prefer spelling out the type.

Reply:
OK, I see your point. (Don't feel you _cannot_ lean on type inference though.)